urlac47192b66_ggrandes_bouncer_tgz-pJ8-org_javastack_bouncer_ClusterTestJ8/src/org/javastack/bouncer/KeyGenerator.java

```diff
@@ -0,0 +1,178 @@
+	 * @param algName algorithm name, eg SHA256withRSA
+	 */
+	static X509Certificate generateCertificate(String dn, KeyPair pair, int days, String algName)
+			throws Exception {
+		final PrivateKey privkey = pair.getPrivate();
+		final X509CertInfo info = new X509CertInfo();
+		final Date from = new Date();
+		final Date to = new Date(from.getTime() + days * 86400000L);
+		final CertificateValidity interval = new CertificateValidity(from, to);
+		final int sn = (int) ((System.currentTimeMillis() / 1000) & 0xFFFFFFFF);
+		final X500Name owner = new X500Name(dn);
+
+		final AlgorithmId algo = AlgorithmId.get(algName);
+		info.set(X509CertInfo.VALIDITY, interval);
+		info.set(X509CertInfo.SERIAL_NUMBER, new CertificateSerialNumber(sn));
+		info.set(X509CertInfo.SUBJECT, new CertificateSubjectName(owner));
+		info.set(X509CertInfo.ISSUER, new CertificateIssuerName(owner));
+		info.set(X509CertInfo.KEY, new CertificateX509Key(pair.getPublic()));
+		info.set(X509CertInfo.VERSION, new CertificateVersion(CertificateVersion.V3));
+		info.set(X509CertInfo.ALGORITHM_ID, new CertificateAlgorithmId(algo));
+
+		// Extensions
+		final CertificateExtensions ext = new CertificateExtensions();
+		ext.set(BasicConstraintsExtension.NAME, new BasicConstraintsExtension(Boolean.TRUE, true, 0)); // Critical|isCA|pathLen
+		ext.set(SubjectKeyIdentifierExtension.NAME,
+				new SubjectKeyIdentifierExtension(new KeyIdentifier(pair.getPublic()).getIdentifier()));
+		ext.set(AuthorityKeyIdentifierExtension.NAME, new AuthorityKeyIdentifierExtension(new KeyIdentifier(
+				pair.getPublic()), null, null));
+		// Extended Key Usage Extension
+		final Vector<ObjectIdentifier> ekue = new Vector<ObjectIdentifier>();
+		ekue.add(new ObjectIdentifier(new int[] {
+				1, 3, 6, 1, 5, 5, 7, 3, 1
+		})); // Server
+		ekue.add(new ObjectIdentifier(new int[] {
+				1, 3, 6, 1, 5, 5, 7, 3, 2
+		})); // Client
+		ext.set(ExtendedKeyUsageExtension.NAME, new ExtendedKeyUsageExtension(Boolean.FALSE, ekue));
+		info.set(X509CertInfo.EXTENSIONS, ext);
+
+		// Sign the X.509
+		final X509CertImpl cert = new X509CertImpl(info);
+		cert.sign(privkey, algo.getName());
+		return cert;
+	}
+
+	static void writeBufferBase64(OutputStream out, byte[] bufIn) throws IOException {
+		final byte[] buf = DatatypeConverter.printBase64Binary(bufIn).getBytes();
+		final int BLOCK_SIZE = 64;
+		for (int i = 0; i < buf.length; i += BLOCK_SIZE) {
+			out.write(buf, i, Math.min(BLOCK_SIZE, buf.length - i));
+			out.write('\r');
+			out.write('\n');
+		}
+	}
+
+	static void writeCertificate(OutputStream out, X509Certificate crt) throws Exception {
+		final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+		baos.write("-----BEGIN CERTIFICATE-----\r\n".getBytes());
+		writeBufferBase64(baos, crt.getEncoded());
+		baos.write("-----END CERTIFICATE-----\r\n".getBytes());
+		out.write(baos.toByteArray());
+		out.flush();
+		out.close();
+		System.out.println(baos.toString());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 158:165, 169:176.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.